packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -0,0 +1,250 @@
+   */
+  async allProducts() {
+    return cachedResult(
+      this.log,
+      this.redis,
+      'listProducts',
+      this.cacheTtlSeconds,
+      () => this.fetchAllProducts()
+    );
+  }
+
+  /**
+   * Fetches all plans from stripe and returns them.
+   *
+   * Use `allPlans` below to use the cached-enhanced version.
+   *
+   * @returns {Promise<AbbrevPlan[]>} All the plans.
+   */
+  async fetchAllPlans() {
+    const plans = [];
+    for await (const item of this.stripe.plans.list({
+      active: true,
+      expand: ['data.product'],
+    })) {
+      // FIXME: Should probably error here if we can't set a product id/name.
+      let product_id, product_name, product_metadata;
+      if (item.product && typeof item.product !== 'string') {
+        product_id = item.product.id;
+        product_name = item.product.name;
+        product_metadata = item.product.metadata;
+      }
+
+      plans.push({
+        plan_id: item.id,
+        plan_name: item.nickname,
+        plan_metadata: item.metadata,
+        product_id,
+        product_name,
+        product_metadata,
+        interval: item.interval,
+        amount: item.amount,
+        currency: item.currency,
+      });
+    }
+    return plans;
+  }
+
+  /**
+   * Fetches all plans from stripe and returns them.
+   *
+   * Uses Redis caching if configured.
+   *
+   * @returns {Promise<AbbrevPlan[]>} All the plans.
+   */
+  async allPlans() {
+    return cachedResult(
+      this.log,
+      this.redis,
+      'listPlans',
+      this.cacheTtlSeconds,
+      () => this.fetchAllPlans()
+    );
+  }
+
+  /**
+   * Verify that the `planId` is a valid upgrade for this `productId`.
+   *
+   * Throws an error if its an invalid upgrade.
+   *
+   * @param {string} productId
+   * @param {string} planId
+   * @returns {Promise<void>}
+   */
+  async verifyPlanUpgradeForSubscription(productId, planId) {
+    const allPlans = await this.allPlans();
+    const currentPlan = allPlans
+      .filter(plan => plan.product_id === productId)
+      .shift();
+
+    const newPlan = allPlans.filter(plan => plan.plan_id === planId).shift();
+    if (!newPlan || !currentPlan) {
+      throw error.unknownSubscriptionPlan();
+    }
+    if (
+      !subhub.validateProductUpgrade(
+        currentPlan.product_metadata,
+        newPlan.product_metadata
+      )
+    ) {
+      throw error.invalidPlanUpgrade();
+    }
+  }
+
+  /**
+   * Change a subscription to the new plan.
+   *
+   * Note that this call does not verify its a valid upgrade, the
+   * `verifyPlanUpgradeForSubscription` should be done first to
+   * validate this is an appropraite change for tier use.
+   *
+   * @param {Subscription['id']} subscriptionId
+   * @param {Plan['id']} planId
+   */
+  async changeSubscriptionPlan(subscriptionId, planId) {
+    const subscription = await this.stripe.subscriptions.retrieve(
+      subscriptionId
+    );
+    if (subscription.items.data[0].plan.id === planId) {
```

Comment:
Is the whole chain of properties guaranteed to be there?

Reply:
A subscription must have at least one plan associated with it, so yes, its guaranteed that the first element of data must exist. This code as-is is not compatible with multi-product should a subscription have multiple plans with it though, as we'll need to know which plan in the subscription should be updated at that point. But that's a later feature. 😉